### PR TITLE
RocketEngine fix for selecting top learners

### LINKED ISSUE #262: PipelineInference\RocketEngine bug fix for taking top k items.
There is a bug in the RocketEngine GetTopLearners() method which potentially allows for duplicated pipelines to show up in the results. Bug found and diagnosed by @yaeldekel. Fix should be easy.

More detail: The logic in GetTopLearners() looped through two arrays, and compared the items in the second array to the items in the first, replacing them if some criterion was met (had higher weights). The way it was written could allow for a learner to replace more than one item in the first list, leading to two or more copies of the learner in our Top K array (which is a bug).

### PR DESCRIPTION
**Pull request by george-microsoft**
Fixes #262, changing the way the top K learners are selected. Now uses a LINQ statement for streamlined code. 

Also adds the full namespace path to the Newtonsoft.Json.Formatting object, to avoid confusion with an object that exists in the Microsoft.ML namespace.

### CODE REVIEW

File: src/Microsoft.ML.PipelineInference/AutoMlEngines/RocketEngine.cs [hunk 1]
```diff
@@ -196,21 +196,10 @@ private void SampleHyperparameters(RecipeInference.SuggestedRecipe.SuggestedLear
...
-                        if (weights[i] > topKTuples[j].Item1)
-                            topKTuples[j] = new Tuple<double, int>(weights[i], i);
-                }
-            }
-
-            return topKTuples.Select(t => AvailableLearners[t.Item2]).ToArray();
+            return weights.Select((w, i) => new { Weight = w, Index = i })
+                .OrderByDescending(x => x.Weight)
+                .Take(_topK)
+                .Select(t=>AvailableLearners[t.Index]).ToArray();
```
Comment (TomFinley) [inline src/Microsoft.ML.PipelineInference/AutoMlEngines/RocketEngine.cs]: >t=>Av [](start = 24, length = 5)

ctrl-k-d to keep formatting up to date and whatnot

File: src/Microsoft.ML.PipelineInference/AutoMlEngines/RocketEngine.cs [hunk 2]
```diff
@@ -196,21 +196,10 @@ private void SampleHyperparameters(RecipeInference.SuggestedRecipe.SuggestedLear
...
-
-            for (int i = 0; i < weights.Length; i++)
-            {
-                if (i < _topK)
-                    topKTuples[i] = new Tuple<double, int>(weights[i], i);
-                else
-                {
-                    for (int j = 0; j < topKTuples.Length; j++)
-                        if (weights[i] > topKTuples[j].Item1)
-                            topKTuples[j] = new Tuple<double, int>(weights[i], i);
```
Comment (TomFinley) [inline src/Microsoft.ML.PipelineInference/AutoMlEngines/RocketEngine.cs]: Oh geez. :)

File: src/Microsoft.ML.PipelineInference/AutoMlEngines/RocketEngine.cs (hunk 1 above)
Comment (george-microsoft) [reply src/Microsoft.ML.PipelineInference/AutoMlEngines/RocketEngine.cs]: Loading the file into Visual Studio and pressing ctrl+k+d has no effect.  (I'm guessing that is already the "preferred" indentation?)

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.PipelineInference/AutoMlEngines/RocketEngine.cs]: Tom probably talks about t=>AvailableLearners and want it to look like t => AvailableLearners.
My VS btw put spaces, maybe setting for that on your machine is disabled.

---
In reply to: [191921117](https://github.com/dotnet/machinelearning/pull/270#discussion_r191921117) [](ancestors = 191921117)

**Pull request merged**